src/tlo/analysis/life_expectancy.py

```diff
@@ -0,0 +1,305 @@
+    """
+    for a single run, estimate life expectancy for males and females
+    return: pd.Series
+    """
+
+    estimated_life_expectancy = pd.Series(dtype=float)
+
+    # first age-group is 0, then 1-4, 5-9, 10-14 etc. 22 categories in total
+    level_0_values = person_years_at_risk.index.get_level_values('age_group').unique()
+    interval_width = pd.Series([1] + [4] + [5] * 18, index=level_0_values)
+
+    fraction_of_last_age_survived = 0.5
+    number_age_groups = 20
+
+    # separate male and female data
+    for sex in ['M', 'F']:
+        person_years_by_sex = person_years_at_risk.xs(key=sex, level='sex')
+        number_of_deaths_by_sex = number_of_deaths_in_interval.xs(key=sex, level='sex')
+
+        death_rate_in_interval = number_of_deaths_by_sex / person_years_by_sex
+        # if no deaths or person-years, convert nan to 0
+        death_rate_in_interval = death_rate_in_interval.fillna(0)
+        # if no deaths in age 90+, set death rate equal to value in age 85-89
+        if death_rate_in_interval[number_age_groups - 1] == 0:
+            death_rate_in_interval[number_age_groups - 1] = death_rate_in_interval[number_age_groups - 2]
+
+        # Calculate the probability of dying in the interval
+        condition = number_of_deaths_by_sex > (
+            person_years_by_sex / interval_width / fraction_of_last_age_survived)
+        probability_of_dying_in_interval = pd.Series(index=number_of_deaths_by_sex.index, dtype=float)
+        probability_of_dying_in_interval[condition] = 1
+        probability_of_dying_in_interval[~condition] = interval_width * death_rate_in_interval / (
+            1 + interval_width * (1 - fraction_of_last_age_survived) * death_rate_in_interval)
+        # all those surviving to final interval die during this interval
+        probability_of_dying_in_interval[number_age_groups - 1] = 1
+
+        # number_alive_at_start_of_interval
+        # keep dtype as float in case using aggregated outputs
+        # note range stops BEFORE the specified number
+        number_alive_at_start_of_interval = pd.Series(index=range(number_age_groups), dtype=float)
+        number_alive_at_start_of_interval[0] = 100_000  # hypothetical cohort
+        for i in range(1, number_age_groups):
+            number_alive_at_start_of_interval[i] = (1 - probability_of_dying_in_interval[i - 1]) * \
+                                                   number_alive_at_start_of_interval[i - 1]
+
+        # number_dying_in_interval
+        number_dying_in_interval = pd.Series(index=range(number_age_groups), dtype=float)
+        for i in range(0, number_age_groups - 1):
+            number_dying_in_interval[i] = number_alive_at_start_of_interval[i] - number_alive_at_start_of_interval[
+                i + 1]
+
+        number_dying_in_interval[number_age_groups - 1] = number_alive_at_start_of_interval[number_age_groups - 1]
+
+        # person-years lived in interval
```

Comment:
Yes I was following the standard life-table approach. I can use the person-years within each interval instead of population size at the start of the interval to generate the death rates within intervals, but using the hypothetical cohort for the actual life expectancy calculation makes it more stable as running this on small populations will often lead to errors. If the number of person-years in any interval is 0 then it won't run. This is especially an issue for older populations in scenarios where we have high mortality. 

Also, I think generally a population size of 5000 is required for reliable estimates of life expectancy. As many people (including me) might want to run this function on small test populations, I went for the "safer" version using the hypothetical cohort. If we feel strongly that this shouldn't be allowed to run on small populations, then I'm happy to adjust the code.